crates/exex/src/lib.rs

```diff
@@ -0,0 +1,140 @@
+use futures_util::{future::BoxFuture, FutureExt};
+use reth_db::database::Database;
+use reth_primitives::BlockNumber;
+use reth_provider::{
+    AccountReader, BlockReaderIdExt, CanonStateNotification, CanonStateNotifications,
+    CanonStateSubscriptions, ChainSpecProvider, ChangeSetReader, DatabaseProviderFactory,
+    EvmEnvProvider, StateProviderFactory,
+};
+use std::future::Future;
+use tokio::sync::watch;
+
+pub struct ExExHandle {
+    /// Meta information about the ExEx.
+    meta: ExExMeta,
+    /// A watch channel that will receive the height of the last block processed by the ExEx.
+    ///
+    /// ExEx should guarantee that it will not require all earlier blocks in the future, meaning
+    /// that Reth is allowed to prune them.
+    ///
+    /// On reorgs, it's possible for the height to go down.
+    finished_height: watch::Receiver<BlockNumber>,
+}
+
+/// The main trait for the custom ExEx.
+///
+/// The ExEx should listen to the canonical state notifications and process them, updating the
+/// `finished_height` in [ExExHandle] accordingly.
+pub trait IntoExEx<DB: Database, P: FullProvider<DB>> {
+    /// Spawns the ExEx.
+    ///
+    /// Returns a future that should be polled to advance the ExEx and a handle to the ExEx.
+    fn spawn_exex(
+        self,
+        provider: P,
+        canon_state_notifications: CanonStateNotifications,
+    ) -> (BoxFuture<'static, ()>, ExExHandle);
+}
+
+// == optional ==
+#[derive(Debug, Clone)]
+pub struct ExExMeta {
+    /// The identifier of the ExEx. It will appear in logs and metrics.
+    id: String,
+    /// The height from which the ExEx should start processing blocks. Earlier blocks will not be
+    /// sent to the ExEx.
+    ///
+    /// In case when the ExEx needs all blocks from the genesis, it should be set to `0`.
+    from_height: BlockNumber,
+}
+
+/// The managed version of the [IntoExEx] trait that covers the common use case and simplifies the
+/// usage.
+pub trait ExEx<DB: Database, P: FullProvider<DB>>: Send + Sync + Unpin + 'static {
+    /// Initializes the ExEx. Called once when the ExEx is spawned. ExEx should persist the
+    /// dependencies such as the provider to re-use them in the future when processing state
+    /// notifications.
+    ///
+    /// Returns the meta information about the ExEx.
+    fn init(&self, provider: P) -> ExExMeta;
+
+    /// Processes the state notification. Once processed, the ExEx will not ever receive the same
+    /// notification again, and Reth is allowed to prune the blocks from notification.
+    ///
+    /// [Result::Err] does not stop the ExEx, and used only for observability purposes. The ExEx
```

Comment:
```suggestion
    /// [Result::Err] does not stop the ExEx, and is used only for observability purposes. The ExEx
```

should this be the intended behavior? for now, let's go with it, but i wonder if this is helpful or not